File: app/src/main/java/android/hcpl/be/filmtrack/model/Roll.java
package android.hcpl.be.filmtrack.model;

import java.io.Serializable;

/**
 * Created by hcpl on 1/08/15.
 */
public class Roll implements Serializable{

    // TODO
//    private long id;

    private String type;

    private int speed, frames;

    public Roll() {
    }

    public Roll(String type, int speed, int frames) {
        this.type = type;
        this.speed = speed;
        this.frames = frames;
    }

    public String getType() {
        return type;
    }

    public void setType(String type) {
        this.type = type;
    }

    public int getSpeed() {
        return speed;
    }

    public void setSpeed(int speed) {
        this.speed = speed;
    }

    public int getFrames() {
        return frames;
    }

    public void setFrames(int frames) {
        this.frames = frames;
    }

    @Override
    public String toString() {
        return new StringBuilder(type).append(" @ ISO ").append(speed).append(" # ").append(String.valueOf(frames)).toString();
    }

    @Override
    public boolean equals(Object o) {
        if (this == o) return true;
        if (!(o instanceof Roll)) return false;

        Roll roll = (Roll) o;

        if (getSpeed() != roll.getSpeed()) return false;
        if (getFrames() != roll.getFrames()) return false;
        return getType().equals(roll.getType());

    }

    @Override
    public int hashCode() {
        int result = getType().hashCode();
        result = 31 * result + getSpeed();
        result = 31 * result + getFrames();
        return result;
    }
}

File: app/src/main/java/be/hcpl/android/filmtag/model/Roll.kt
package be.hcpl.android.filmtag.model

import androidx.annotation.Keep
import java.io.Serializable
import java.util.ArrayList

@Keep
data class Roll(
    var id: Long = -1,
    var type: String? = null,
    var speed: Int = 200,
    var frames: Int = 36,
    var notes: String? = null,
    var isDeveloped: Boolean = false,
    var tags: List<String> = ArrayList(),
) : Serializable {

    init {
        id = if (id > 0) id else System.currentTimeMillis() // generates unique ID for all objects created
    }

    override fun toString(): String {
        return StringBuilder(type.orEmpty()).append(" @ ISO ").append(speed).append(" # ").append(frames.toString()).toString()
    }

    override fun equals(o: Any?): Boolean {
        return if (this === o) true
        else (o as? Roll)?.id == id

    }

    override fun hashCode(): Int {
        return (id xor id.ushr(32)).toInt()
    }
}
